src/core.rs

```diff
@@ -330,10 +342,32 @@ impl Hypercore {
         })
     }
 
+    #[cfg(feature = "replication")]
+    /// Subscribe to core events relevant to replication
+    pub fn event_subscribe(
+        &self,
+    ) -> tokio::sync::broadcast::Receiver<crate::replication::events::Event> {
+        self.events.channel.subscribe()
+    }
+
+    /// Check if core has the block at the given `index` locally
+    #[instrument(ret, skip(self))]
+    pub fn has(&self, index: u64) -> bool {
+        self.bitfield.get(index)
+    }
+
     /// Read value at given index, if any.
     #[instrument(err, skip(self))]
     pub async fn get(&mut self, index: u64) -> Result<Option<Vec<u8>>, HypercoreError> {
         if !self.bitfield.get(index) {
+            #[cfg(feature = "replication")]
+            // if not in this core, try to get over network
+            {
+                let mut rx = self.events.send_on_get(index);
+                tokio::spawn(async move {
```

Comment:
I think that would make sense, but should probably be configurable like the [js get method is](https://github.com/datrs/hyperswarm-rs/pull/3), to have `wait` and `timeout` parameter. However changing this methods signature would be a breaking change. So I didn't want to do it in this PR.

The default behavior in js seems to be `wait = true` but `timeout = 0`. Which effectively does the same thing here, it notifies the replicator that we want to get a block, but does not wait on it. Relevant JS [here](https://github.com/holepunchto/hypercore/blob/7dcd252d281dfc3f2dd62260d27573901791c4b6/index.js#L897-L908)

I propose adding a separate `Hypercore::get_opt(index: u64, opts: GetOpts) -> Future<Result<...>>` but doing so in a separate PR.

Reply:
Ah I see. So with this code the behavior would look like:

```rust
if core.get(1337).is_none() {  // triggers download in background
// some time later, do we wait for an event?
core.get(1337) -> Some(bytes) // got downloaded!
}
```

It feels like it'd be better to decouple `try get + trigger download` into separate things. Like maybe `get()` would return `None` always until you trigger a download via `download(index|range).await`.

Then code would look like

```rust
if core.get(1337).is_none() {
  core.download(1337).await?;
  return core.get(1337);
}

Reply:
The current behavior matches JS, I think diverging from that could be confusing to users. I'm unsure about adding a `Hypercore::download` method at this time since it would **require** the `replication` feature and would be harder to do in a async runtime agnostic way.

Reply:
I was thinking about this before. There is actually no need for the `spawn`. This should just send the event that we tried to get. It is up to the subscriber for how to handle it. But in the case of the replicator implementation, the behavior is the same as JS.